src/main/java/net/dries007/tfc/common/blocks/devices/SheetPileBlock.java

```diff
@@ -257,15 +276,52 @@ public boolean onDestroyedByPlayer(BlockState state, Level level, BlockPos pos,
     }
 
     @Override
-    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder)
-    {
-        super.createBlockStateDefinition(builder.add(PROPERTIES));
+    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
+        super.createBlockStateDefinition(builder.add(PROPERTIES).add(FACING).add(MIRROR));
     }
 
     @Override
     @SuppressWarnings("deprecation")
-    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context)
-    {
+    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
         return shapeCache.get(state);
     }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState rotate(BlockState state, Rotation rot) {
+        return switch (rot) {
+            case CLOCKWISE_90 -> state.setValue(NORTH, state.getValue(WEST))
+                .setValue(EAST, state.getValue(NORTH))
+                .setValue(SOUTH, state.getValue(EAST))
+                .setValue(WEST, state.getValue(SOUTH))
+                .setValue(FACING, rot.rotate(state.getValue(FACING)));
+            case CLOCKWISE_180 -> state.setValue(NORTH, state.getValue(SOUTH))
+                .setValue(EAST, state.getValue(WEST))
+                .setValue(SOUTH, state.getValue(NORTH))
+                .setValue(WEST, state.getValue(EAST))
+                .setValue(FACING, rot.rotate(state.getValue(FACING)));
+            case COUNTERCLOCKWISE_90 -> state.setValue(NORTH, state.getValue(EAST))
+                .setValue(EAST, state.getValue(SOUTH))
+                .setValue(SOUTH, state.getValue(WEST))
+                .setValue(WEST, state.getValue(NORTH))
+                .setValue(FACING, rot.rotate(state.getValue(FACING)));
+            default -> state;
+        };
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState mirror(BlockState state, Mirror mirror) {
+        return switch (mirror) {
+            case LEFT_RIGHT -> state.setValue(NORTH, state.getValue(SOUTH))
+                .setValue(SOUTH, state.getValue(NORTH))
+                .setValue(FACING, mirror.mirror(state.getValue(FACING)))
+                .cycle(MIRROR);
+            case FRONT_BACK -> state.setValue(EAST, state.getValue(WEST))
+                .setValue(WEST, state.getValue(EAST))
+                .setValue(FACING, mirror.mirror(state.getValue(FACING)))
+                .cycle(MIRROR);
+            default -> state;
+        };
+    }
```

Comment:
For these methods (ones which rotate `NORTH`, `SOUTH`, `EAST`, and `WEST`, and on blocks that inherit `DirectionPropertyBlock`, (`SheetPileBlock`, `CreepingPlantBlock`, also for `KelpTreeBlock`), could you delegate these to a static method in `DirectionPropertyBlock` ? As it's a fair amount of duplicated code otherwise.